guardian/typerighter
# Discard draft changes

## What does this change?



This PR adds a 'Discard unpublished changes' button when a Live rule has unpublished changes. It replaces the 'Has unpublished changes' warning text with a button that reverts the changes.

Currently, when a user makes changes to a Live rule, a warning appears saying the rule 'Has unpublished changes'. These changes are saved to the draft rule. However, it's not possible for the user to see which changes have been made since the latest live rule, nor is it possible to revert these changes. Say a cat walks over your keyboard as you have a rule open - this button will help you undo the work of their paws!

When a user clicks to 'Discard unpublished changes', the draft rule will be updated with properties from the latest live rule, reverting it to a previous version of the rule.

The only property of the draft rule that will be different from the previous state of the draft rule is the `updatedAt` time. We can't retrieve the `updatedAt` time of the draft rule's previous state at the time of publication - this field will have been overwritten by making changes to the draft rule and this property won't necessary match the `createdAt` time of the latest live rule. So instead the `updatedAt` property of the reverted rule will be set to match the `updatedAt` time of the latest live rule. A mock `updatedAt` time is used in the test.

Trello card is [here](https://trello.com/c/fzHRxgtw/1526-add-a-discard-draft-changes-butto).

## How to test



The test can be run within `sbt` with `testOnly *RuleManagerSpec -- -z "revertDraftRule"`

To test this as a user, do the following:
- publish a draft rule
- make changes to the live rule - the 'Discard changes' button should appear in the 'RULE STATUS' bar - next to the 'Has unpublished changes' warning
- click the 'Discard changes button' - the rule should revert to the most recently published state of the live rule

## Images



### Before changes to live rule
<img width="494" alt="discard changes - before changes" src="https://github.com/guardian/typerighter/assets/7883129/169393ee-98f2-4265-a880-141fe88ff5cb">

### After changes to live rule
<img width="493" alt="discard changes - after changes" src="https://github.com/guardian/typerighter/assets/7883129/e881d1de-f0de-4d1e-aecd-476974fa2e71">
